crates/proto/src/serialize/binary/encoder.rs

```diff
@@ -209,8 +209,7 @@ impl<'a> BinEncoder<'a> {
     pub fn trim(&mut self) {
         let offset = self.offset;
         self.buffer.truncate(offset);
-        self.name_pointers
-            .retain(|&(start, end)| start < offset && end <= offset);
+        self.name_pointers.retain(|&(start, _)| start < offset);
```

Comment:
so the primary fix is this removing `end <= offset`, correct?

We need to be careful that we don't reintroduce this bug: https://github.com/bluejekyll/trust-dns/pull/341/files#diff-f3f8112c5398e636353cd27fbf9d6c61R118

I think the test coverage has this, but I just want to point there to make sure we don't have a potential issue there.

Reply:
I see this comment resolved, but just like to clarity that the primary fix is not about `end <= offset` but changing the entire struct that instead of storing the start and end positions (that represent pointers), the fix stores the actual full labels, because in the original code, the `pointers` [will get removed for labels ending with a pointer](https://github.com/bluejekyll/trust-dns/blob/main/crates/proto/src/rr/domain/name.rs#L623). We set the offset back and then trim whatever afterwards. I have thought of some optimization, still using pointers but this requires some kind of de-reference  on labels with pointer to match appropriately. And it seems de-referencing might even outweigh this solution that introduces a certain amount of space complexity due to storing full labels as opposed to pointer positions. With this said, I am working on testing my software more thoroughly on a more wider user base to see if I can obtain some benchmark results for comparison.